{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Glycolysis\n",
    "\n",
    "To construct a model of glycolysis, first we import  __masspy__ . "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from mass import (\n",
    "    io, MassMetabolite, MassReaction, MassModel, Simulation, plot_simulation, qcqa_model)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Model Construction \n",
    "\n",
    "The first step of creating a model of glycolysis is to define the MassModel object. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "glycolysis = MassModel(\"Glycolysis\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Metabolites\n",
    "\n",
    "The next step is to define all of the metabolites in glycolysis using identifiers and the names found in the BiGG Database (http://bigg.ucsd.edu/). In order to ensure our model is physiologically accurate, it is important to provide the __formula__ keyword with a string representing the chemical formula for each metabolite, and the  __charge__ keyword with an integer representing the metabolite's ionic charge (Note that neutrally charged metabolites are provided with 0). To indicate that the cytosol is the cellular compartment in which glycolysis occurs, the string \"c\" is provided to the __compartment__ keyword. \n",
    "\n",
    "In this model of glycolysis, there are 20 metabolites inside the cytosol compartment. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Glycolytic Intermediates\n",
    "glc__D_c = MassMetabolite(\n",
    "    id=\"glc__D_c\",\n",
    "    name=\"D-Glucose\",\n",
    "    formula=\"C6H12O6\",\n",
    "    charge=0,\n",
    "    compartment=\"c\")\n",
    "g6p_c = MassMetabolite(\n",
    "    id=\"g6p_c\",\n",
    "    name=\"D-Glucose 6-phosphate\",\n",
    "    formula=\"C6H11O9P\",\n",
    "    charge=-2,\n",
    "    compartment=\"c\")\n",
    "f6p_c = MassMetabolite(\n",
    "    id=\"f6p_c\",\n",
    "    name=\"D-Fructose 6-phosphate\",\n",
    "    formula=\"C6H11O9P\",\n",
    "    charge=-2,\n",
    "    compartment=\"c\")\n",
    "fdp_c = MassMetabolite(\n",
    "    id=\"fdp_c\",\n",
    "    name=\"D-Fructose 1,6-bisphosphate\",\n",
    "    formula=\"C6H10O12P2\",\n",
    "    charge=-4,\n",
    "    compartment=\"c\")\n",
    "dhap_c = MassMetabolite(\n",
    "    id=\"dhap_c\",\n",
    "    name=\"Dihydroxyacetone phosphate\",\n",
    "    formula=\"C3H5O6P\",\n",
    "    charge=-2,\n",
    "    compartment=\"c\")\n",
    "g3p_c = MassMetabolite(\n",
    "    id=\"g3p_c\",\n",
    "    name=\"Glyceraldehyde 3-phosphate\",\n",
    "    formula=\"C3H5O6P\",\n",
    "    charge=-2,\n",
    "    compartment=\"c\")\n",
    "_13dpg_c = MassMetabolite(\n",
    "    id=\"_13dpg_c\",\n",
    "    name=\"3-Phospho-D-glyceroyl phosphate\",\n",
    "    formula=\"C3H4O10P2\",\n",
    "    charge=-4,\n",
    "    compartment=\"c\")\n",
    "_3pg_c = MassMetabolite(\n",
    "    id=\"_3pg_c\",\n",
    "    name=\"3-Phospho-D-glycerate\",\n",
    "    formula=\"C3H4O7P\",\n",
    "    charge=-3,\n",
    "    compartment=\"c\")\n",
    "_2pg_c = MassMetabolite(\n",
    "    id=\"_2pg_c\",\n",
    "    name=\"D-Glycerate 2-phosphate\",\n",
    "    formula=\"C3H4O7P\",\n",
    "    charge=-3,\n",
    "    compartment=\"c\")\n",
    "pep_c = MassMetabolite(\n",
    "    id=\"pep_c\",\n",
    "    name=\"Phosphoenolpyruvate\",\n",
    "    formula=\"C3H2O6P\",\n",
    "    charge=-3,\n",
    "    compartment=\"c\")\n",
    "pyr_c = MassMetabolite(\n",
    "    id=\"pyr_c\",\n",
    "    name=\"Pyruvate\",\n",
    "    formula=\"C3H3O3\",\n",
    "    charge=-1,\n",
    "    compartment=\"c\")\n",
    "lac__L_c = MassMetabolite(\n",
    "    id=\"lac__L_c\",\n",
    "    name=\"L-Lactate\",\n",
    "    formula=\"C3H5O3\",\n",
    "    charge=-1,\n",
    "    compartment=\"c\")\n",
    "# Cofactors\n",
    "nad_c = MassMetabolite(\n",
    "    id=\"nad_c\",\n",
    "    name=\"Nicotinamide adenine dinucleotide\",\n",
    "    formula=\"C21H26N7O14P2\",\n",
    "    charge=-1,\n",
    "    compartment=\"c\")\n",
    "nadh_c = MassMetabolite(\n",
    "    id=\"nadh_c\",\n",
    "    name=\"Nicotinamide adenine dinucleotide - reduced\",\n",
    "    formula=\"C21H27N7O14P2\",\n",
    "    charge=-2,\n",
    "    compartment=\"c\")\n",
    "atp_c = MassMetabolite(\n",
    "    id=\"atp_c\",\n",
    "    name=\"ATP\",\n",
    "    formula=\"C10H12N5O13P3\",\n",
    "    charge=-4,\n",
    "    compartment=\"c\")\n",
    "adp_c = MassMetabolite(\n",
    "    id=\"adp_c\",\n",
    "    name=\"ADP\",\n",
    "    formula=\"C10H12N5O10P2\",\n",
    "    charge=-3,\n",
    "    compartment=\"c\")\n",
    "amp_c = MassMetabolite(\n",
    "    id=\"amp_c\",\n",
    "    name=\"AMP\",\n",
    "    formula=\"C10H12N5O7P\",\n",
    "    charge=-2,\n",
    "    compartment=\"c\")\n",
    "h_c = MassMetabolite(\n",
    "    id=\"h_c\",\n",
    "    name=\"H+\",\n",
    "    formula=\"H\",\n",
    "    charge=1,\n",
    "    compartment=\"c\")\n",
    "pi_c = MassMetabolite(\n",
    "    id=\"pi_c\",\n",
    "    name=\"Phosphate\",\n",
    "    formula=\"HPO4\",\n",
    "    charge=-2,\n",
    "    compartment=\"c\")\n",
    "h2o_c = MassMetabolite(\n",
    "    id=\"h2o_c\",\n",
    "    name=\"H2O\",\n",
    "    formula=\"H2O\",\n",
    "    charge=0,\n",
    "    compartment=\"c\")\n",
    "\n",
    "metabolite_list = [glc__D_c, g6p_c, f6p_c, fdp_c, dhap_c, \n",
    "                   g3p_c, _13dpg_c, _3pg_c, _2pg_c, pep_c,\n",
    "                   pyr_c, lac__L_c, nad_c, nadh_c, atp_c, \n",
    "                   adp_c, amp_c, h_c, pi_c, h2o_c]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After generating the metabolites, all metabolites are added to the model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "glycolysis.add_metabolites(metabolite_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reactions\n",
    "\n",
    "Once all of the MassMetabolite objects have been created for each metabolite in glycolysis, the next step is to define all of the reactions that occur inside the cytosol compartment using BiGG idenfifiers and reaction names. To make this model useful for integration with other models, it is important to provide the string \"Glycolysis\" to the __subsystem__ keyword. By providing the subsystem, the reactions can be easily obtained even when integrated with a significantly larger model. \n",
    "\n",
    "Note that after the creation of each MassReaction object, the metabolites are added to the reaction using a dictionary where keys are the metabolites and values are the stoichiometric coefficients (reactants have negative coefficients, products have positive ones). \n",
    "\n",
    "In this model of glycolysis, there are 14 reactions occuring inside the cytosol."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "HEX1 = MassReaction(\n",
    "    id=\"HEX1\",\n",
    "    name=\"Hexokinase (D-glucose:ATP)\",\n",
    "    subsystem=\"Glycolysis\")\n",
    "HEX1.add_metabolites({\n",
    "    glc__D_c: -1,\n",
    "    atp_c: -1,\n",
    "    adp_c: 1,\n",
    "    g6p_c: 1,\n",
    "    h_c: 1})\n",
    "\n",
    "PGI = MassReaction(\n",
    "    id=\"PGI\",\n",
    "    name=\"Glucose-6-phosphate isomerase\",\n",
    "    subsystem=\"Glycolysis\")\n",
    "PGI.add_metabolites({\n",
    "    g6p_c: -1,\n",
    "    f6p_c: 1})\n",
    "\n",
    "PFK = MassReaction(\n",
    "    id=\"PFK\",\n",
    "    name=\"Phosphofructokinase\",\n",
    "    subsystem=\"Glycolysis\")\n",
    "PFK.add_metabolites({\n",
    "    f6p_c: -1,\n",
    "    atp_c: -1,\n",
    "    fdp_c: 1,\n",
    "    adp_c: 1,\n",
    "    h_c: 1})\n",
    "\n",
    "FBA = MassReaction(\n",
    "    id=\"FBA\",\n",
    "    name=\"Fructose-bisphosphate aldolase\",\n",
    "    subsystem=\"Glyolysis\")\n",
    "FBA.add_metabolites({\n",
    "    fdp_c: -1,\n",
    "    dhap_c: 1,\n",
    "    g3p_c: 1})\n",
    "\n",
    "TPI = MassReaction(\n",
    "    id=\"TPI\",\n",
    "    name=\"Triose-phosphate isomerase\",\n",
    "    subsystem=\"Glycolysis\")\n",
    "TPI.add_metabolites({\n",
    "    dhap_c: -1,\n",
    "    g3p_c: 1})\n",
    "\n",
    "GAPD = MassReaction(\n",
    "    id=\"GAPD\",\n",
    "    name=\"Glyceraldehyde-3-phosphate dehydrogenase\",\n",
    "    subsystem=\"Glycolysis\")\n",
    "GAPD.add_metabolites({\n",
    "    g3p_c: -1,\n",
    "    nad_c: -1,\n",
    "    pi_c: -1,\n",
    "    _13dpg_c: 1,\n",
    "    h_c: 1,\n",
    "    nadh_c: 1})\n",
    "\n",
    "PGK = MassReaction(\n",
    "    id=\"PGK\",\n",
    "    name=\"Phosphoglycerate kinase\",\n",
    "    subsystem=\"Glycolysis\")\n",
    "PGK.add_metabolites({\n",
    "    _13dpg_c: -1,\n",
    "    adp_c: -1,\n",
    "    _3pg_c: 1,\n",
    "    atp_c: 1})\n",
    "\n",
    "PGM = MassReaction(\n",
    "    id=\"PGM\",\n",
    "    name=\"Phosphoglycerate mutase\",\n",
    "    subsystem=\"Glycolysis\")\n",
    "PGM.add_metabolites({\n",
    "    _3pg_c: -1,\n",
    "    _2pg_c: 1})\n",
    "\n",
    "ENO = MassReaction(\n",
    "    id=\"ENO\",\n",
    "    name=\"Enolase\",\n",
    "    subsystem=\"Glycolysis\")\n",
    "ENO.add_metabolites({\n",
    "    _2pg_c: -1,\n",
    "    h2o_c: 1,\n",
    "    pep_c: 1})\n",
    "\n",
    "PYK = MassReaction(\n",
    "    id=\"PYK\",\n",
    "    name=\"Pyruvate kinase\",\n",
    "    subsystem=\"Glycolysis\")\n",
    "PYK.add_metabolites({\n",
    "    pep_c: -1,\n",
    "    h_c: -1,\n",
    "    adp_c: -1,\n",
    "    atp_c: 1,\n",
    "    pyr_c: 1})\n",
    "\n",
    "LDH_L = MassReaction(\n",
    "    id=\"LDH_L\",\n",
    "    name=\"L-lactate dehydrogenase\",\n",
    "    subsystem=\"Glycolysis\")\n",
    "LDH_L.add_metabolites({\n",
    "    h_c: -1,\n",
    "    nadh_c: -1,\n",
    "    pyr_c: -1,\n",
    "    lac__L_c: 1,\n",
    "    nad_c: 1})\n",
    "\n",
    "reaction_list = [HEX1, PGI, PFK, FBA, TPI, \n",
    "                 GAPD, PGK, PGM, ENO, PYK, \n",
    "                 LDH_L]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In addition to the reactions, the ATP maintenence reaction (ATPM), the Adenylate kinase reaction (ADK1), and the NADH demand reaction (DM_NADH) must also be included in the model. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "ADK1 = MassReaction(\n",
    "    id=\"ADK1\",\n",
    "    name=\"Adenylate kinase\",\n",
    "    subsystem=\"Misc.\")\n",
    "ADK1.add_metabolites({\n",
    "    adp_c: -2,\n",
    "    amp_c: 1,\n",
    "    atp_c: 1})\n",
    "\n",
    "ATPM = MassReaction(\n",
    "    id=\"ATPM\",\n",
    "    name=\"ATP maintenance requirement\",\n",
    "    subsystem=\"Pseudo\", \n",
    "    reversible=False)\n",
    "ATPM.add_metabolites({\n",
    "    atp_c: -1,\n",
    "    h2o_c: -1,\n",
    "    adp_c: 1,\n",
    "    h_c: 1,\n",
    "    pi_c: 1})\n",
    "DM_nadh = MassReaction(\n",
    "    id=\"DM_nadh\",\n",
    "    name=\"Demand NADH\",\n",
    "    subsystem=\"Pseudo\",\n",
    "    reversible=False)\n",
    "DM_nadh.add_metabolites({\n",
    "    nadh_c: -1,\n",
    "    nad_c: 1,\n",
    "    h_c: 1})\n",
    "\n",
    "reaction_list += [ADK1, ATPM, DM_nadh]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After generating the reactions, all reactions are added to the glycolysis model. Adding the MassReaction objects will also add their associated MassMetabolite objects if they have not already been added to the model. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "glycolysis.add_reactions(reaction_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exchange reactions\n",
    "\n",
    "After generating the reactions, the next step is to add the exchange reactions and fix the concentrations of the external metabolites of the system. This can easily be done using the __MassModel.add_exchange__ method. With the generation of the exchange reactions, the character \"e\" is used to defined the \"Extracellular\" compartment. Therefore, it is recommended to add the definition to the __MassModel.compartment__ attribute. \n",
    "\n",
    "In this model of glycolysis, there are 8 exchange reactions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <table>\n",
       "                <tr>\n",
       "                    <td><strong>Name</strong></td><td>Glycolysis</td>\n",
       "                </tr><tr>\n",
       "                    <td><strong>Memory address</strong></td><td>0x01232fc5f8</td>\n",
       "                </tr><tr>\n",
       "                    <td><strong>Stoichiometric Matrix</strong></td>\n",
       "                    <td>20x22</td>\n",
       "                </tr><tr>\n",
       "                    <td><strong>Matrix Rank</strong></td>\n",
       "                    <td>19</td>\n",
       "                </tr><tr>\n",
       "                    <td><strong>Matrix Type</strong></td>\n",
       "                    <td>dense, float64</td>\n",
       "                </tr><tr>\n",
       "                    <td><strong>Number of Metabolites</strong></td>\n",
       "                    <td>20</td>\n",
       "                </tr><tr>\n",
       "                    <td><strong>Number of Reactions</strong></td>\n",
       "                    <td>22</td>\n",
       "                </tr><tr>\n",
       "                    <td><strong>Number of Initial Conditions</strong></td>\n",
       "                    <td>0</td>\n",
       "                </tr><tr>\n",
       "                    <td><strong>Number of Forward Rate Constants</strong></td>\n",
       "                    <td>0</td>\n",
       "                </tr><tr>\n",
       "                    <td><strong>Number of Equilibrium Constants</strong></td>\n",
       "                    <td>5</td>\n",
       "                </tr><tr>\n",
       "                    <td><strong>Number of Irreversible Reactions</strong></td>\n",
       "                    <td>5</td>\n",
       "                </tr><tr>\n",
       "                    <td><strong>Number of Exchanges</strong></td>\n",
       "                    <td>8</td>\n",
       "                </tr><tr>\n",
       "                    <td><strong>Number of Fixed Concentrations</strong></td>\n",
       "                    <td>7</td>\n",
       "                </tr><tr>\n",
       "                    <td><strong>Number of Custom Rates</strong></td>\n",
       "                    <td>0</td>\n",
       "                </tr><tr>\n",
       "                    <td><strong>Number of Genes</strong></td>\n",
       "                    <td>0</td>\n",
       "                </tr><tr>\n",
       "                    <td><strong>Number of Enzymes</strong></td>\n",
       "                    <td>0</td>\n",
       "                </tr><tr>\n",
       "                    <td><strong>Modules</strong></td>\n",
       "                    <td></br></td>\n",
       "                </tr><tr>\n",
       "                    <td><strong>Compartments</strong></td>\n",
       "                    <td>c</td>\n",
       "                </tr><tr>\n",
       "                    <td><strong>Units</strong></td>\n",
       "                    <td></td>\n",
       "                </tr>\n",
       "            </table>\n",
       "        "
      ],
      "text/plain": [
       "<MassModel Glycolysis at 0x1232fc5f8>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "S_glc__D_e = glycolysis.add_exchange(glc__D_c, exchange_type=\"source\", \n",
    "                                      external_concentration=1)\n",
    "EX_lac__L_e = glycolysis.add_exchange(lac__L_c, exchange_type=\"exchange\", \n",
    "                                      external_concentration=1)\n",
    "EX_pyr_e = glycolysis.add_exchange(pyr_c, exchange_type=\"exchange\", \n",
    "                                      external_concentration=0.06)\n",
    "EX_h_e = glycolysis.add_exchange(h_c, exchange_type=\"exchange\", \n",
    "                                      external_concentration=6.30957e-05)\n",
    "EX_h2o_e = glycolysis.add_exchange(h2o_c, exchange_type=\"exchange\", \n",
    "                                      external_concentration=1)\n",
    "S_amp_e = glycolysis.add_exchange(amp_c, exchange_type=\"source\", \n",
    "                                      external_concentration=1)\n",
    "DM_amp_e = glycolysis.add_exchange(amp_c, exchange_type=\"demand\", \n",
    "                                      external_concentration=1)\n",
    "EX_pi_e = glycolysis.add_exchange(pi_c, exchange_type=\"exchange\", \n",
    "                                 external_concentration=2.5)\n",
    "\n",
    "glycolysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Optional: Ordering of internal species and reactions\n",
    "\n",
    "Sometimes, it is also desirable to reorder the metabolite and reaction objects inside the model to follow the physiology. To reorder the internal species, one can use the method __cobra.DictList.get_by_any__ method with the list of metabolite identifiers in the desirable order. To ensure all metabolites are still present and not forgotten in the model, a QC/QA check is also performed. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cobra import DictList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HEX1</th>\n",
       "      <th>PGI</th>\n",
       "      <th>PFK</th>\n",
       "      <th>FBA</th>\n",
       "      <th>TPI</th>\n",
       "      <th>GAPD</th>\n",
       "      <th>PGK</th>\n",
       "      <th>PGM</th>\n",
       "      <th>ENO</th>\n",
       "      <th>PYK</th>\n",
       "      <th>...</th>\n",
       "      <th>ADK1</th>\n",
       "      <th>EX_pyr_e</th>\n",
       "      <th>EX_lac__L_e</th>\n",
       "      <th>ATPM</th>\n",
       "      <th>DM_nadh</th>\n",
       "      <th>S_glc__D_e</th>\n",
       "      <th>S_amp_e</th>\n",
       "      <th>EX_h_e</th>\n",
       "      <th>EX_h2o_e</th>\n",
       "      <th>EX_pi_e</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>glc__D_c</th>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g6p_c</th>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f6p_c</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fdp_c</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dhap_c</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g3p_c</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>_13dpg_c</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>_3pg_c</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>_2pg_c</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pep_c</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pyr_c</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lac__L_c</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nad_c</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nadh_c</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>amp_c</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>adp_c</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atp_c</th>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pi_c</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>h_c</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>h2o_c</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          HEX1  PGI  PFK  FBA  TPI  GAPD  PGK  PGM  ENO  PYK   ...     ADK1  \\\n",
       "glc__D_c  -1.0  0.0  0.0  0.0  0.0   0.0  0.0  0.0  0.0  0.0   ...      0.0   \n",
       "g6p_c      1.0 -1.0  0.0  0.0  0.0   0.0  0.0  0.0  0.0  0.0   ...      0.0   \n",
       "f6p_c      0.0  1.0 -1.0  0.0  0.0   0.0  0.0  0.0  0.0  0.0   ...      0.0   \n",
       "fdp_c      0.0  0.0  1.0 -1.0  0.0   0.0  0.0  0.0  0.0  0.0   ...      0.0   \n",
       "dhap_c     0.0  0.0  0.0  1.0 -1.0   0.0  0.0  0.0  0.0  0.0   ...      0.0   \n",
       "g3p_c      0.0  0.0  0.0  1.0  1.0  -1.0  0.0  0.0  0.0  0.0   ...      0.0   \n",
       "_13dpg_c   0.0  0.0  0.0  0.0  0.0   1.0 -1.0  0.0  0.0  0.0   ...      0.0   \n",
       "_3pg_c     0.0  0.0  0.0  0.0  0.0   0.0  1.0 -1.0  0.0  0.0   ...      0.0   \n",
       "_2pg_c     0.0  0.0  0.0  0.0  0.0   0.0  0.0  1.0 -1.0  0.0   ...      0.0   \n",
       "pep_c      0.0  0.0  0.0  0.0  0.0   0.0  0.0  0.0  1.0 -1.0   ...      0.0   \n",
       "pyr_c      0.0  0.0  0.0  0.0  0.0   0.0  0.0  0.0  0.0  1.0   ...      0.0   \n",
       "lac__L_c   0.0  0.0  0.0  0.0  0.0   0.0  0.0  0.0  0.0  0.0   ...      0.0   \n",
       "nad_c      0.0  0.0  0.0  0.0  0.0  -1.0  0.0  0.0  0.0  0.0   ...      0.0   \n",
       "nadh_c     0.0  0.0  0.0  0.0  0.0   1.0  0.0  0.0  0.0  0.0   ...      0.0   \n",
       "amp_c      0.0  0.0  0.0  0.0  0.0   0.0  0.0  0.0  0.0  0.0   ...      1.0   \n",
       "adp_c      1.0  0.0  1.0  0.0  0.0   0.0 -1.0  0.0  0.0 -1.0   ...     -2.0   \n",
       "atp_c     -1.0  0.0 -1.0  0.0  0.0   0.0  1.0  0.0  0.0  1.0   ...      1.0   \n",
       "pi_c       0.0  0.0  0.0  0.0  0.0  -1.0  0.0  0.0  0.0  0.0   ...      0.0   \n",
       "h_c        1.0  0.0  1.0  0.0  0.0   1.0  0.0  0.0  0.0 -1.0   ...      0.0   \n",
       "h2o_c      0.0  0.0  0.0  0.0  0.0   0.0  0.0  0.0  1.0  0.0   ...      0.0   \n",
       "\n",
       "          EX_pyr_e  EX_lac__L_e  ATPM  DM_nadh  S_glc__D_e  S_amp_e  EX_h_e  \\\n",
       "glc__D_c       0.0          0.0   0.0      0.0         1.0      0.0     0.0   \n",
       "g6p_c          0.0          0.0   0.0      0.0         0.0      0.0     0.0   \n",
       "f6p_c          0.0          0.0   0.0      0.0         0.0      0.0     0.0   \n",
       "fdp_c          0.0          0.0   0.0      0.0         0.0      0.0     0.0   \n",
       "dhap_c         0.0          0.0   0.0      0.0         0.0      0.0     0.0   \n",
       "g3p_c          0.0          0.0   0.0      0.0         0.0      0.0     0.0   \n",
       "_13dpg_c       0.0          0.0   0.0      0.0         0.0      0.0     0.0   \n",
       "_3pg_c         0.0          0.0   0.0      0.0         0.0      0.0     0.0   \n",
       "_2pg_c         0.0          0.0   0.0      0.0         0.0      0.0     0.0   \n",
       "pep_c          0.0          0.0   0.0      0.0         0.0      0.0     0.0   \n",
       "pyr_c         -1.0          0.0   0.0      0.0         0.0      0.0     0.0   \n",
       "lac__L_c       0.0         -1.0   0.0      0.0         0.0      0.0     0.0   \n",
       "nad_c          0.0          0.0   0.0      1.0         0.0      0.0     0.0   \n",
       "nadh_c         0.0          0.0   0.0     -1.0         0.0      0.0     0.0   \n",
       "amp_c          0.0          0.0   0.0      0.0         0.0      1.0     0.0   \n",
       "adp_c          0.0          0.0   1.0      0.0         0.0      0.0     0.0   \n",
       "atp_c          0.0          0.0  -1.0      0.0         0.0      0.0     0.0   \n",
       "pi_c           0.0          0.0   1.0      0.0         0.0      0.0     0.0   \n",
       "h_c            0.0          0.0   1.0      1.0         0.0      0.0    -1.0   \n",
       "h2o_c          0.0          0.0  -1.0      0.0         0.0      0.0     0.0   \n",
       "\n",
       "          EX_h2o_e  EX_pi_e  \n",
       "glc__D_c       0.0      0.0  \n",
       "g6p_c          0.0      0.0  \n",
       "f6p_c          0.0      0.0  \n",
       "fdp_c          0.0      0.0  \n",
       "dhap_c         0.0      0.0  \n",
       "g3p_c          0.0      0.0  \n",
       "_13dpg_c       0.0      0.0  \n",
       "_3pg_c         0.0      0.0  \n",
       "_2pg_c         0.0      0.0  \n",
       "pep_c          0.0      0.0  \n",
       "pyr_c          0.0      0.0  \n",
       "lac__L_c       0.0      0.0  \n",
       "nad_c          0.0      0.0  \n",
       "nadh_c         0.0      0.0  \n",
       "amp_c          0.0      0.0  \n",
       "adp_c          0.0      0.0  \n",
       "atp_c          0.0      0.0  \n",
       "pi_c           0.0     -1.0  \n",
       "h_c            0.0      0.0  \n",
       "h2o_c         -1.0      0.0  \n",
       "\n",
       "[20 rows x 22 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_metabolite_order = [\"glc__D_c\", \"g6p_c\", \"f6p_c\", \"fdp_c\", \"dhap_c\", \n",
    "                        \"g3p_c\", \"_13dpg_c\", \"_3pg_c\", \"_2pg_c\", \"pep_c\", \n",
    "                        \"pyr_c\", \"lac__L_c\", \"nad_c\", \"nadh_c\", \"amp_c\", \n",
    "                        \"adp_c\", \"atp_c\", \"pi_c\", \"h_c\", \"h2o_c\"]\n",
    "\n",
    "if len(glycolysis.metabolites) == len(new_metabolite_order):\n",
    "    glycolysis.metabolites = DictList(glycolysis.metabolites.get_by_any(new_metabolite_order))\n",
    "    \n",
    "new_reaction_order = [\"HEX1\", \"PGI\", \"PFK\", \"FBA\", \"TPI\", \n",
    "                      \"GAPD\", \"PGK\", \"PGM\", \"ENO\", \"PYK\", \n",
    "                      \"LDH_L\", \"DM_amp_e\", \"ADK1\", \"EX_pyr_e\", \n",
    "                      \"EX_lac__L_e\", \"ATPM\", \"DM_nadh\", \"S_glc__D_e\", \n",
    "                      \"S_amp_e\", \"EX_h_e\", \"EX_h2o_e\",\"EX_pi_e\"]\n",
    "\n",
    "if len(glycolysis.reactions) == len(new_reaction_order):\n",
    "    glycolysis.reactions = DictList(glycolysis.reactions.get_by_any(new_reaction_order))\n",
    "    \n",
    "glycolysis.update_S(matrix_type=\"DataFrame\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Model Parameterization\n",
    "### Steady State fluxes\n",
    "Steady state fluxes can be computed as a summation of the MinSpan pathway vectors. \n",
    "\n",
    "Pathways are obtained using the minspan package (https://github.com/SBRG/minspan). \n",
    "\n",
    "Using these pathways and literature sources, independent fluxes can be defined in order to calculate the steady state flux vector. For this model, flux and concentration data are obtained from the historical human red blood cell (RBC) metabolic model (Heinrich 1977, Jamshidi 2002, Joshi 1988, Joshi 1990, Joshi 1989-II). \n",
    "\n",
    "From the literature, it is known that:\n",
    "\n",
    "1. A typical uptake rate of the RBC of glucose is about 1.12 mM/hr. \n",
    "2. Based on experimental data, the steady state load on NADH is set at 20% of the glucose uptake rate, or $0.2 * 1.12 = 0.244. \n",
    "3. The synthesis rate of AMP is measured to be 0.014 mM/hr. \n",
    "\n",
    "With these pathays and numerical values, the steady state flux vector can be computed as the weighted sum of the corresponding basis vectors. The steady state flux vector is computed as an inner product:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HEX1: 1.12\n",
      "PGI: 1.12\n",
      "PFK: 1.12\n",
      "FBA: 1.12\n",
      "TPI: 1.12\n",
      "GAPD: 2.24\n",
      "PGK: 2.24\n",
      "PGM: 2.24\n",
      "ENO: 2.24\n",
      "PYK: 2.24\n",
      "LDH_L: 2.016\n",
      "DM_amp_e: 0.014\n",
      "ADK1: 0.0\n",
      "EX_pyr_e: 0.22400000000000003\n",
      "EX_lac__L_e: 2.016\n",
      "ATPM: 2.24\n",
      "DM_nadh: 0.22400000000000003\n",
      "S_glc__D_e: 1.12\n",
      "S_amp_e: 0.014\n",
      "EX_h_e: 2.688\n",
      "EX_h2o_e: 0.0\n",
      "EX_pi_e: 0.0\n"
     ]
    }
   ],
   "source": [
    "minspan_paths = [\n",
    "    [1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 0, 0, 0, 2, 2, 0, 1, 0, 2, 0, 0],\n",
    "    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0,-1, 0, 0, 1,-1, 0, 1, 0, 0, 2, 0, 0],\n",
    "    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0]]\n",
    "\n",
    "glyc_ss_fluxes = glycolysis.compute_steady_state_fluxes(\n",
    "                    minspan_paths, \n",
    "                    {S_glc__D_e: 1.12, DM_nadh: .2*1.12, DM_amp_e: 0.014},\n",
    "                    update_reactions=True);\n",
    "\n",
    "for rxn in glycolysis.reactions:\n",
    "    print(\"%s: %s\" % (rxn.id, rxn.v))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Initial Conditions\n",
    "\n",
    "Once the network has been built, the concentrations can be added to the metabolites. The cytosolic metabolite concentrations are added to each individual metabolite using the __MassMetabolite.initial_condition__ or __MassMetabolite.ic__  setter methods. Once added, the __MassModel.set_initial_conditions__ method is used to add the initial conditions into the MassModel object itself. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "glc__D_c.ic = 1.0\n",
    "g6p_c.ic = 0.0486\n",
    "f6p_c.ic = 0.0198\n",
    "fdp_c.ic = 0.0146\n",
    "g3p_c.ic = 0.00728\n",
    "dhap_c.ic = 0.16\n",
    "_13dpg_c.ic = 0.000243\n",
    "_3pg_c.ic = 0.0773\n",
    "_2pg_c.ic = 0.0113\n",
    "pep_c.ic = 0.017\n",
    "pyr_c.ic = 0.060301\n",
    "lac__L_c.ic = 1.36\n",
    "atp_c.ic = 1.6\n",
    "adp_c.ic = 0.29\n",
    "amp_c.ic = 0.0867281\n",
    "h_c.ic = 0.0000899757\n",
    "nad_c.ic = 0.0589\n",
    "nadh_c.ic = 0.0301\n",
    "pi_c.ic = 2.5\n",
    "h2o_c.ic = 1.0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Equilibirum Constants\n",
    "\n",
    "After adding initial conditions and steady state fluxes, the equilibrium constants  are defined using the __MassReaction.equilibrium_constant__ or __MassReaction.Keq__ setter methods. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cytosolic reactions\n",
    "HEX1.Keq = 850\n",
    "PGI.Keq = 0.41\n",
    "PFK.Keq = 310\n",
    "FBA.Keq = 0.082\n",
    "TPI.Keq = 0.0571429\n",
    "GAPD.Keq = 0.0179\n",
    "PGK.Keq = 1800\n",
    "PGM.Keq = 0.147059\n",
    "ENO.Keq = 1.69492\n",
    "PYK.Keq = 363000\n",
    "LDH_L.Keq = 26300\n",
    "ADK1.Keq = 1.65\n",
    "\n",
    "#Excahnges\n",
    "EX_lac__L_e.Keq = 1\n",
    "EX_pyr_e.Keq = 1\n",
    "EX_h_e.Keq = 1\n",
    "EX_h2o_e.Keq = 1\n",
    "EX_pi_e.Keq = 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculation of PERCs\n",
    "By defining the equilibrium constant and steady state parameters, the values of the pseudo rate constants (PERCs) can be calculated and added to the model using the __MassModel.calc_PERCS__ method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Cannot calculate the PERC for reaction 'HEX1' because values for ['atp_c', 'adp_c', 'g6p_c', 'glc__D_c'] not defined.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-14-762176015694>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mpercs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mglycolysis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcalculate_PERCs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mupdate_reactions\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m;\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mrxn_kf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mperc\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mpercs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"%s: %.3f\"\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mrxn_kf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mperc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/github/masspy/mass/core/massmodel.py\u001b[0m in \u001b[0;36mcalculate_PERCs\u001b[0;34m(self, steady_state_fluxes, steady_state_concentrations, at_equilibrium_default, update_reactions)\u001b[0m\n\u001b[1;32m   1573\u001b[0m                 raise ValueError(\"Cannot calculate the PERC for reaction '{0}'\"\n\u001b[1;32m   1574\u001b[0m                                  \u001b[0;34m\" because values for {1} not defined.\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1575\u001b[0;31m                                  .format(reaction.id, str(missing_values)))\n\u001b[0m\u001b[1;32m   1576\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1577\u001b[0m             \u001b[0;31m# Substitute values into rate equation\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Cannot calculate the PERC for reaction 'HEX1' because values for ['atp_c', 'adp_c', 'g6p_c', 'glc__D_c'] not defined."
     ]
    }
   ],
   "source": [
    "percs = glycolysis.calculate_PERCs(update_reactions=True);\n",
    "for rxn_kf, perc in percs.items():\n",
    "    print(\"%s: %.3f\" % (rxn_kf, perc))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From literature, it is known that the rate of the phosphate uptake is 0.26/hr. Therefore, the PERC for the phosphate exchange is updated manually:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "EX_pi_e.kf = 0.26"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## QC/QA Model\n",
    "\n",
    "Before we simulate the model, it is important to ensure that the model is elementally balanced, and that the model can simulate. Therefore, the __qcqa_model__ method is used to provide a report on the model quality and indicate whether simulation is possible and if not, what parameters and/or initial conditions are missing. \n",
    "\n",
    "__Note:__ Typically exchange and demand reactions are necessarily elementally balanced and therefore should not generally be a cause for concern. Because the NADH reaction generates electrons for processes that use redox potential, it is expected to not be balanced."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "╒══════════════════════════════════════════════╕\n",
      "│ MODEL ID: Glycolysis                         │\n",
      "│ SIMULATABLE: False;                          │\n",
      "│ NUMERICAL CONSISTENCY: True                  │\n",
      "╞══════════════════════════════════════════════╡\n",
      "│ ============================================ │\n",
      "│             MISSING PARAMETERS               │\n",
      "│ ============================================ │\n",
      "│ Reaction Parameters                          │\n",
      "│ ---------------------                        │\n",
      "│ HEX1: kf; kr                                 │\n",
      "│ PGI: kf; kr                                  │\n",
      "│ PFK: kf; kr                                  │\n",
      "│ FBA: kf; kr                                  │\n",
      "│ TPI: kf; kr                                  │\n",
      "│ GAPD: kf; kr                                 │\n",
      "│ PGK: kf; kr                                  │\n",
      "│ PGM: kf; kr                                  │\n",
      "│ ENO: kf; kr                                  │\n",
      "│ PYK: kf; kr                                  │\n",
      "│ LDH_L: kf; kr                                │\n",
      "│ DM_amp_e: kf                                 │\n",
      "│ ADK1: kf; kr                                 │\n",
      "│ EX_pyr_e: kf; kr                             │\n",
      "│ EX_lac__L_e: kf; kr                          │\n",
      "│ ATPM: kf                                     │\n",
      "│ DM_nadh: kf                                  │\n",
      "│ S_glc__D_e: kf                               │\n",
      "│ S_amp_e: kf                                  │\n",
      "│ EX_h_e: kf; kr                               │\n",
      "│ EX_h2o_e: kf; kr                             │\n",
      "│ EX_pi_e: kf; kr                              │\n",
      "│ ============================================ │\n",
      "├──────────────────────────────────────────────┤\n",
      "│ ============================================ │\n",
      "│           MISSING CONCENTRATIONS             │\n",
      "│ ============================================ │\n",
      "│ Initial Conditions                           │\n",
      "│ --------------------                         │\n",
      "│ glc__D_c                                     │\n",
      "│ g6p_c                                        │\n",
      "│ f6p_c                                        │\n",
      "│ fdp_c                                        │\n",
      "│ dhap_c                                       │\n",
      "│ g3p_c                                        │\n",
      "│ _13dpg_c                                     │\n",
      "│ _3pg_c                                       │\n",
      "│ _2pg_c                                       │\n",
      "│ pep_c                                        │\n",
      "│ pyr_c                                        │\n",
      "│ lac__L_c                                     │\n",
      "│ nad_c                                        │\n",
      "│ nadh_c                                       │\n",
      "│ amp_c                                        │\n",
      "│ adp_c                                        │\n",
      "│ atp_c                                        │\n",
      "│ pi_c                                         │\n",
      "│ h_c                                          │\n",
      "│ h2o_c                                        │\n",
      "│ ============================================ │\n",
      "├──────────────────────────────────────────────┤\n",
      "│ ============================================ │\n",
      "│             CONSISTENCY CHECKS               │\n",
      "│ ============================================ │\n",
      "│ Elemental                                    │\n",
      "│ ---------------------------------            │\n",
      "│ DM_nadh: {charge: 2.0} unbalanced            │\n",
      "│ ============================================ │\n",
      "╘══════════════════════════════════════════════╛\n"
     ]
    }
   ],
   "source": [
    "qcqa_model(glycolysis, parameters=True, concentrations=True, \n",
    "           fluxes=True, superfluous=True, elemental=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From the results of the QC/QA test, it can be seen that the model can be simulated and is numerically consistent."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Steady State and Model Validation "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To find the steady state of the model, the __Simulation.find_steady_state_model__ method can used, and the \n",
    "__Simulation.update_values__ will update the values in the Simulation object with the steady state values. \n",
    "To validate the steady state of the model, concentration and flux solutions can be plotted using the __plot_simulation__ function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "asdasddas\n",
    "sim = Simulation(glycolysis)\n",
    "\n",
    "t0, tf = (0, 1e6)\n",
    "conc_sol, flux_sol = sim.simulate_model(glycolysis, time=(t0, tf))\n",
    "\n",
    "fig, (ax1, ax2) = plt.subplots(nrows=2, ncols=1, figsize=(11, 8), constrained_layout=True);\n",
    "\n",
    "plot_simulation(conc_sol, ax=ax1, legend=\"right outside\",\n",
    "                plot_function=\"semilogx\", xlabel=\"Time [hr]\",\n",
    "                ylabel=\"Concentrations [mM]\");\n",
    "\n",
    "plot_simulation(flux_sol, ax=ax2, legend=\"right outside\",\n",
    "                plot_function=\"semilogx\", xlabel=\"Time [hr]\",\n",
    "                ylabel=\"Fluxes [mM]\");"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Optional Information: Storing references\n",
    "#### Compartment\n",
    "Because the character \"c\" represents the cytosol compartment, it is recommended to define and set the compartment in the __MassModel.compartments__ attribute."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "glycolysis.compartments = {\"c\": \"Cytosol\"}\n",
    "print(glycolysis.compartments)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Units\n",
    "All of the units for the numerical values used in this model are \"Millimoles\" for amount, \"Liters\" for volume, and \"Hours\" for time. In order to ensure that future users understand the numerical values for model, it is important to define the __MassModel.units__ attribute. Like the __compartment__ attribute, the __units__ attribute is a dictionary and can easily be updated using the same methods.\n",
    "\n",
    "Note that this attribute will NOT track units, but instead acts as a reference for the user and others so that they can perform necessary unit conversions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "glycolysis.units = {'N': 'Millimoles', 'Vol': 'Liters', 'Time': 'Hours'}\n",
    "glycolysis.units"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Literature\n",
    "\n",
    "Because specific sources of literature were used in the building of this model, it is important for others who wish to use this model to be able to access those same literature sources. In general, to preserve information for users, either the __notes__ attribute or the __annotation__ attribute can be used. In this case, the literature source is placed in __notes__. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "glycolysis.notes.update({\"Literature\": \n",
    "    [\"R. Heinrich, S. M. Rapoport, and T. A. Rapoport. Metabolic regulations and mathematical models. Prog. Biophys. Molec. Biol., 32:1–82, 1977.\",\n",
    "     \"A. Joshi. Integrated Metabolic Dynamics in the Human Red Blood Cell. PhD thesis, University of Michigan, 1988.\",\n",
    "     \"N. Jamshidi, S.J. Wiback, and B.O. Palsson. In silico model-driven assess- ment of the effects of single nucleotide polymorphisms (SNPs) on human red blood cell metabolism. Genome Research, 12:1687–1692, 2002.,\"\n",
    "     \"A. Joshi. Integrated Metabolic Dynamics in the Human Red Blood Cell. PhD thesis, University of Michigan, 1988.\",\n",
    "     \"A. Joshi and B.O. Palsson. Metabolic dynamics in the human red cell. Part IV–data prediction and some model computations. J. theor. Biol., 142:69–85, 1990.\",\n",
    "     \"A. Joshi and B.O. Palsson. Metabolic dynamics in the human red cell. Part I I –interactions with the environment. J. theor. Biol., 141:529–545, 1989.\",\n",
    "     \"Prankerd, T. A. J., & Altman, K. I. (1954). A study of the metabolism of phosphorus in mammalian red cells. Biochemical Journal, 58(4), 622–633.\"\n",
    "    ]})\n",
    "glycolysis.notes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Export\n",
    "\n",
    "After validation, the model is ready to be saved. The model can either be exported as a \".json\" file or if the model conforms to SBML standards, as a \".xml\" file using the classes in __mass.io__. To export the model, only the path to the directory and the name of the model need to be specified. Both are shown. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Export using JSON"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "io.json.save_json_model(glycolysis, os.path.abspath(\"../../test-models/json-models/glycolysis.json\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Export using SBML"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# io.sbml.write_sbml_model(glycolysis, os.path.abspath(\"../../test-models/sbml-models/glycolysis.xml\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
